library/psa_crypto.c

```diff
@@ -2840,6 +2842,260 @@ psa_status_t psa_mac_verify_finish( psa_mac_operation_t *operation,
 /* Asymmetric cryptography */
 /****************************************************************/
 
+typedef enum
+{
+    PSA_SIGN_INVALID = 0,
+    PSA_SIGN_HASH = 1,
+    PSA_SIGN_MESSAGE
+} psa_sign_operation_t;
+
+typedef enum
+{
+    PSA_VERIFY_INVALID = 0,
+    PSA_VERIFY_HASH = 1,
+    PSA_VERIFY_MESSAGE
+} psa_verify_operation_t;
+
+static psa_status_t psa_sign_internal( mbedtls_svc_key_id_t key,
+                                       psa_sign_operation_t operation,
+                                       psa_algorithm_t alg,
+                                       const uint8_t * input,
+                                       size_t input_length,
+                                       uint8_t * signature,
+                                       size_t signature_size,
+                                       size_t * signature_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_status_t unlock_status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_slot_t *slot;
+
+    *signature_length = 0;
+
+    if( operation == PSA_SIGN_MESSAGE )
+    {
+        if( ! PSA_ALG_IS_SIGN_MESSAGE( alg ) )
+            return( PSA_ERROR_INVALID_ARGUMENT );
+
+        if ( PSA_ALG_IS_HASH_AND_SIGN( alg ) )
+        {
+            if( ! PSA_ALG_IS_HASH( PSA_ALG_SIGN_GET_HASH( alg ) ) )
+                return( PSA_ERROR_INVALID_ARGUMENT );
+        }
+        /* Curently only hash-then-sign algorithms are supported. */
+        else
+            return( PSA_ERROR_INVALID_ARGUMENT );
+    }
+
+    else if( operation == PSA_SIGN_INVALID )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
+    /* Immediately reject a zero-length signature buffer. This guarantees
+     * that signature must be a valid pointer. (On the other hand, the hash
+     * buffer can in principle be empty since it doesn't actually have
+     * to be a hash.) */
+    if( signature_size == 0 )
+        return( PSA_ERROR_BUFFER_TOO_SMALL );
+
+    status = psa_get_and_lock_key_slot_with_policy(
+                key, &slot,
+                operation == PSA_SIGN_HASH ? PSA_KEY_USAGE_SIGN_HASH :
```

Comment:
Since according to the spec linked by Gilles earlier "a key with permission `PSA_KEY_USAGE_SIGN/VERIFY_HASH` can always be used for signing/verifying a full message too, why defer that to a separate PR?

The way I see it, that would just require a minor modification to the policy checking code, and since you're introducing testing for sign_message / verify_message in this PR, I think it fully makes sense to do the policy aspect in here as well.

Reply:
The scheduling of this PR made me to put it into a separate PR. We would like to finish this PR as soon as possible.